src/private/lidar_impl.cpp

```diff
@@ -0,0 +1,191 @@
+#include <libm2k/lidar.hpp>
+#include <iio.h>
+#include <iostream>
+
+using namespace std;
+using namespace libm2k::contexts;
+using namespace libm2k::utils;
+
+class Lidar::LidarImpl : public ContextImpl {
+public:
+	LidarImpl(std::string uri, iio_context* ctx, std::string name, bool sync) :
+		ContextImpl(uri, ctx, name, sync) {
+
+		pulse_dev = new DeviceGeneric(ctx, "7c700000.axi-pulse-capture");
+		afe_dev	  = new DeviceGeneric(ctx, "ad5627");
+		adc_dev	  = new DeviceIn     (ctx, "axi-ad9094-hpc");
+
+		setAllIIoAttributesToDefaultValues();
+	}
+
+	void initialize() {
+		;
+	}
+    
+	void channelEnableDisable(string channel, bool enable) {
+		if	(channel == "voltage0")
+			adc_dev->enableChannel(0, enable, false);
+		else if (channel == "voltage1")
+			adc_dev->enableChannel(1, enable, false);
+		else if (channel == "voltage2")
+			adc_dev->enableChannel(2, enable, false);
+		else if (channel == "voltage3")
+			adc_dev->enableChannel(3, enable, false);
+		else if (channel == "voltage4")
+			adc_dev->enableChannel(4, enable, false);
+		else
+			return;		    // invalid channel
+	}
+
+	vector<string> enabledChannelsList() {
+		vector<string> enabled_channels;
+		if (adc_dev->isChannelEnabled(0, false))
+			enabled_channels.push_back("voltage0");
+		if (adc_dev->isChannelEnabled(1, false))
+			enabled_channels.push_back("voltage1");
+		if (adc_dev->isChannelEnabled(2, false))
+			enabled_channels.push_back("voltage2");
+		if (adc_dev->isChannelEnabled(3, false))
+			enabled_channels.push_back("voltage3");
+		if (adc_dev->isChannelEnabled(4, false))
+			enabled_channels.push_back("voltage4");
+		return enabled_channels;
+	}
+
+	///// Example usage for reading multiple channels at the same time,
+	// lidar->channelEnableDisable("voltage0", true);
+	// lidar->channelEnableDisable("voltage1", true);
+	// lidar->channelEnableDisable("voltage2", true);
+	// lidar->channelEnableDisable("voltage3", true);
+	// auto values = lidar->readChannels(512);
+
+	// for (auto itr = values.begin(); itr != values.end(); ++itr) {
+	//     printf("%s\n", itr->first.c_str());  /* Print channel name. */
+	//     for (auto v = itr->second.begin(); v != itr->second.end(); ++v)
+	//	   printf("%d \n", *v);		    /* Print all channel samples.
+	//     printf("\n");
+	// }
+	map<string, vector<int8_t>> readChannels(unsigned int nb_samples) {
+		vector<string> enabled_channels = enabledChannelsList();
+		unsigned int enabled_channels_count = enabled_channels.size();
+	
+		int8_t* start = (int8_t*)adc_dev->getSamplesRawInterleavedVoid(nb_samples);
+		int8_t* sample;
+		int8_t* end   = start + (nb_samples * enabled_channels_count);
+		std::ptrdiff_t step = 1 * enabled_channels_count; // one byte/sample for Lidar
+
+		std::vector<int8_t> samples;
+		std::map<string, vector<int8_t>> result;
+
+		// For every channel that is enabled at the time of buffer refill..
+		for (vector<string>::size_type i = 0; i != enabled_channels_count - 1; i++) {
```

Comment:
what happens here if `enabled_channels_count == 0`?

Reply:
Channel4 is enabled by default at initialization, so at least one channel is always enabled. Unless...c**p, the user might play with the thing and disable all channels, just to see what happens. Good call!

Ok, two possibilities here, as we've discussed before but never reached a consensus. Return an empty result or throw an exception?

Reply:
In cases like this ( or the one we discussed above) my opinion is that the library should throw an exception.
In this particular scenario the following line should throw because no channel is enabled: 
`int8_t* start = (int8_t*)adc_dev->getSamplesRawInterleavedVoid(nb_samples);`
( https://github.com/analogdevicesinc/libm2k/blob/master/src/utils/private/buffer_impl.cpp#L384 )

Reply:
@AlexandraTrifan, confirmed. Thanks! This is a non-issue, then.